staging/src/k8s.io/apiextensions-apiserver/pkg/controller/apiapproval/apiapproval_controller.go

```diff
@@ -0,0 +1,264 @@
+	switch approvalState {
+	case apihelpers.APIApprovalInvalid:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "InvalidAnnotation",
+			Message: reason,
+		}
+	case apihelpers.APIApprovalMissing:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "MissingAnnotation",
+			Message: reason,
+		}
+	case apihelpers.APIApproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "ApprovedAnnotation",
+			Message: reason,
+		}
+	case apihelpers.APIUnapproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "UnapprovedAnnotation",
+			Message: reason,
+		}
+	default:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionUnknown,
+			Reason:  "UnknownAnnotation",
+			Message: reason,
+		}
+	}
+}
+
+func (c *KubeAPIApprovedConditionController) sync(key string) error {
+	inCustomResourceDefinition, err := c.crdLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// avoid repeated calculation for the same annotation
+	protectionAnnotationValue := inCustomResourceDefinition.Annotations[v1beta1.KubeAPIApprovedAnnotation]
+	c.lastSeenProtectedAnnotationLock.Lock()
+	lastSeen, seenBefore := c.lastSeenProtectedAnnotation[inCustomResourceDefinition.Name]
+	c.lastSeenProtectedAnnotationLock.Unlock()
+	if seenBefore && protectionAnnotationValue == lastSeen {
+		return nil
+	}
+
+	// check old condition
+	cond := calculateCondition(inCustomResourceDefinition)
+	if cond == nil {
+		// because group is immutable, if we have no condition now, we have no need to remove a condition.
+		return nil
+	}
+	old := apiextensions.FindCRDCondition(inCustomResourceDefinition, apiextensions.KubeAPIApproved)
+
+	// don't attempt a write if all the condition details are the same
+	if old != nil && old.Status == cond.Status && old.Reason == cond.Reason && old.Message == cond.Message {
+		// no need to update annotation because we took no action.
+		return nil
+	}
+
+	// update condition
+	crd := inCustomResourceDefinition.DeepCopy()
+	apiextensions.SetCRDCondition(crd, *cond)
+
+	_, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+	if apierrors.IsNotFound(err) || apierrors.IsConflict(err) {
+		// deleted or changed in the meantime, we'll get called again
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// store annotation in order to avoid repeated updates for the same annotation (and potential
+	// fights of API server in HA environments).
```

Comment:
> Yeah apiserver fights are problematic if you do what I said in the prior comment. But there's no way to tell the difference between an apiserver and a user.

That seems like overkill for an informational condition.  If someone twiddles this, they're only lying to themselves.  I'd prefer to avoid the risk the apiserver fights and take the chance that a cluster-admin intentionally deceives themselves.  Also, that cluster-admin needs therapy.